mbtiles.go

```diff
@@ -0,0 +1,292 @@
+package main
+
+import (
+	"bytes"
+	"compress/gzip"
+	"compress/zlib"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+type TileFormat uint8
+
+const (
+	UNKNOWN TileFormat = iota
+	GZIP               // encoding = gzip
+	ZLIB               // encoding = deflate
+	PNG
+	JPG
+	PBF
+	WEBP
+)
+
+var TileFormatStr = map[TileFormat]string{
+	PNG:  "png",
+	JPG:  "jpg",
+	PBF:  "pbf",
+	WEBP: "webp",
+}
+
+var TileContentType = map[TileFormat]string{
+	PNG:  "image/png",
+	JPG:  "image/jpeg",
+	PBF:  "application/x-protobuf", // Content-Encoding header must be gzip
+	WEBP: "image/webp",
+}
+
+type Mbtiles struct {
+	filename           string
+	db                 *sql.DB
+	metadata           map[string]interface{}
+	tileformat         TileFormat // tile format: PNG, JPG, PBF
+	timestamp          time.Time  // timestamp of file, for cache control headers
+	hasUTFGrid         bool
+	utfgridCompression TileFormat
+	hasUTFGridData     bool
+}
+
+// Creates a new Mbtiles instance.
+// Connection is closed by runtime on application termination or by calling .Close() method.
+func NewMbtiles(filename string) (*Mbtiles, error) {
+	_, id := filepath.Split(filename)
+	id = strings.Split(id, ".")[0]
+
+	db, err := sql.Open("sqlite3", filename)
```

Comment:
@ka7eh compared to prior version, I am deliberately choosing not to do a `defer thing.Close()` and instead letting that just happen when the server shuts down, because all database connections are kept open the entire time the server is running.